### PR TITLE
Simprints pr rebased

### PR DESCRIPTION
**Pull request by alxndrsn**

### CODE REVIEW

File: static/js/controllers/contacts.js [hunk 1]
```diff
@@ -219,58 +227,23 @@ var SP_ID_MASK = 0xFFFFF8;
...
-          })
-          .catch(function(err) {
-            $scope.error = true;
-            $scope.loading = false;
-            $scope.appending = false;
-            $log.error('Error searching for contacts by simprints ID', err);
+        Simprints.identify().then(function(identities) {
+          identities.forEach(function(identity) {
+            // Tier from TIER_1 (best) to TIER_5 (worst)
+            identity.tierNumber = Number.parseInt(identity.tier.split('_')[1]);
```
Comment (alxndrsn) [inline static/js/controllers/contacts.js]: Could conversion to a number be done within `Simprints` service?

File: static/js/controllers/contacts.js [hunk 2]
```diff
@@ -219,58 +227,23 @@ var SP_ID_MASK = 0xFFFFF8;
...
           });
+          identities = identities.filter(function(identity) {
+            // TIER_1, TIER_2, and TIER_3 are considered a match
+            // TIER_4, and TIER_5 are considered no match
+            // https://sites.google.com/simprints.com/simprints-for-developers/custom-integrations/tiers
+            return identity.tierNumber < 4;
```
Comment (alxndrsn) [inline static/js/controllers/contacts.js]: Interesting!  Probably explains my comment above on the CSS, but are we happy that displaying 3 stars to users for a five star match is appropriate?  If they are also seeing these stars e.g. on the simprints hardware device or in their android app, this may cause confusion.

Comment (alxndrsn) [inline static/js/controllers/contacts.js]: Also it would seem appropriate to do this filtering inside the `Simprints` service.

File: static/js/enketo/widgets/simprints.js [hunk 3]
```diff
@@ -51,22 +32,28 @@ define( function( require, exports, module ) {
...
+        var $input = $el.find( 'input' );
+        $input.hide();
+        var angularServices = angular.element( document.body ).injector();
+        var $translate = angularServices.get( '$translate' );
+        var service = angularServices.get( 'Simprints' );
+
+        if ( !service.enabled() ) {
+            $translate('simprints.disabled').then(function(label) {
+                $el.append( '<p>' + label + '</p>' );
+            });
```
Comment (alxndrsn) [inline static/js/enketo/widgets/simprints.js]: I'd be happier with consistent code formatting within a single file.  It's more important to me that we maintain the enketo style for all boilerplate within our custom widgets, so if you're happier with medic-standard formatting within widget bodies and enketo-style for boilerplate then please continue.

File: static/js/enketo/widgets/simprints.js [hunk 4]
```diff
@@ -51,22 +32,28 @@ define( function( require, exports, module ) {
...
             return;
         }
 
-        /* jslint bitwise: true */
-        var simprintsInputId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER) & SP_ID_MASK;
-        $el.find( 'input' ).attr( 'data-simprints-reg', simprintsInputId );
-        $el.append( '<button class="btn btn-primary" onclick="medicmobile_android.simprints_reg(' + simprintsInputId + ')"><i class="fa fa-bluetooth"> Get new Simprints ID</button>' );
+        $el.on('click', '.btn.simprints-register', function() {
+            service.register().then(function(response) {
+                $input.val(response.id).trigger('change');
```
Comment (alxndrsn) [inline static/js/enketo/widgets/simprints.js]: I'm not sure that cacheing jquery objects for use in event listeners is generally a good thing.  Any ideas on this?  Currently I try to avoid it, but if it's ok then I'll be doing it more.

File: static/js/enketo/widgets/simprints.js [hunk 5]
```diff
@@ -51,22 +32,28 @@ define( function( require, exports, module ) {
...
+            });
+        });
+
+        $translate('simprints.register').then(function(label) {
+            $el.append( '<div><a class="btn btn-default simprints-register"><img src="static/img/simprints.png" width="20" height="20"/> ' + label + '</a></div>' );
```
Comment (alxndrsn) [inline static/js/enketo/widgets/simprints.js]: AFAICT this is prone to injection in the translation.  If we care, this can be prevented by constructing the jquery element more explicitly.

File: static/js/services/generate-search-requests.js [hunk 6]
```diff
@@ -33,15 +33,15 @@ var _ = require('underscore'),
       };
 
       var getRequestWithMappedKeys = function(view, keys, mapKeysFunc) {
-          if (!keys || keys.length === 0) {
-            return;
+        if (!keys || keys.length === 0) {
```
Comment (alxndrsn) [inline static/js/services/generate-search-requests.js]: This is unrelated?

File: templates/partials/filters/simprints.html [hunk 7]
```diff
@@ -1,6 +1,5 @@
 <span class="simprints-filter" ng-show="simprintsEnabled">
-  <a tabindex="-1" ng-click="triggerSimprintsIdent()" translate-attr-title="Search with Simprints">
-    <input type="hidden" name="simprints-idents"/>
-    <i class="fa fa-bluetooth"></i>
+  <a tabindex="-1" ng-click="simprintsIdentify()" translate-attr-title="simprints.search">
+    <img src="static/img/simprints.png" width="20" height="20"/>
```
Comment (alxndrsn) [inline templates/partials/filters/simprints.html]: Seems weird to be putting width and height here as HTML attributes.

File: translations/messages-en.properties [hunk 8]
```diff
@@ -906,4 +906,7 @@ forms.json = JSON
 upload.xform.xml.help = The xform definition in XML.
 upload.xform.meta.help = The meta file in JSON format including the icon, title, context, and so on.
 password.weak = The password is too easy to guess. Include a range of types of characters to increase the score.
-password.length.minimum = The password must be at least {{minimum}} characters long.
\ No newline at end of file
+password.length.minimum = The password must be at least {{minimum}} characters long.
```
Comment (alxndrsn) [inline translations/messages-en.properties]: If you configure your editor to include linebreaks at the end of files, git won't consider the previous-final line of files to have changed when you append more lines later.

File: static/js/controllers/contacts.js [hunk 9]
```diff
@@ -95,9 +96,17 @@ var SP_ID_MASK = 0xFFFFF8;
...
               contacts.splice(homeIndex, 1);
               contacts.unshift(usersHomePlace);
-            } else if (!$scope.filters.search) {
+            } else if (!$scope.filters.search && !$scope.filters.simprintsIdentities) {
               contacts.unshift(usersHomePlace);
             }
+            if ($scope.filters.simprintsIdentities) {
+              contacts.forEach(function(contact) {
+                var identity = $scope.filters.simprintsIdentities.find(function(identity) {
+                  return identity.id === contact.simprints_id;
```
Comment (alxndrsn) [inline static/js/controllers/contacts.js]: Is there a reason to return the `identify()` result as a json object `{ id: id }`, rather than just a string?

File: static/css/inbox.less [hunk 10]
```diff
@@ -1299,6 +1291,21 @@ a.fa:hover {
...
+  &:after {
+    font-family: FontAwesome;
+  }
+  &.tier-1:after {
+    content: '\f005\f005\f005'; /* yes sir, yes sir, three stars full */
+  }
+  &.tier-2:after {
+    content: '\f005\f005\f006'; /* two stars full, one empty */
+  }
+  &.tier-3:after {
```
Comment (garethbowen) [inline static/css/inbox.less]: Discussed below.

File: static/js/controllers/contacts.js (hunk 2 above)
Comment (garethbowen) [inline static/js/controllers/contacts.js]: Good call. I'll move it.

File: static/js/controllers/contacts.js (hunk 9 above)
Comment (garethbowen) [inline static/js/controllers/contacts.js]: Currently the AndroidApi parses the response as JSON and then passes it on to `Simprints.response` function which handles responses of all types. In retrospect this was probably laziness on my part and we should have a `identifyResponse` and `registerResponse` functions so the response can be handled differently. This would also help for the filtering suggested below.

File: static/js/controllers/contacts.js (hunk 2 above)
Comment (garethbowen) [inline static/js/controllers/contacts.js]: I'll also modify the CSS to show two full stars so it matches expectations.

File: static/js/enketo/widgets/simprints.js (hunk 3 above)
Comment (garethbowen) [inline static/js/enketo/widgets/simprints.js]: I'm sorry - which bit is not enketo formatting? It's 4 spaces indented... is it just the spaces around the function parameters?

File: static/js/services/generate-search-requests.js (hunk 6 above)
Comment (garethbowen) [inline static/js/services/generate-search-requests.js]: Yeah, just fixing the indentation while I was in there.

File: translations/messages-en.properties (hunk 8 above)
Comment (garethbowen) [inline translations/messages-en.properties]: I agree. Unfortunately the poeditor service exports these files without the trailing new line, so if we add it then POE keeps removing it anyway.

File: static/js/enketo/widgets/simprints.js (hunk 5 above)
Comment (garethbowen) [inline static/js/enketo/widgets/simprints.js]: We use the ["escape" sanitization strategy](https://angular-translate.github.io/docs/#/guide/19_security) in angular-translate so translation values should not be vulnerable to injection.

File: static/js/controllers/contacts.js (hunk 9 above)
Comment (garethbowen) [inline static/js/controllers/contacts.js]: I've left Android returning JSON but changed the Simprints service to just return the ID.

File: static/js/enketo/widgets/simprints.js (hunk 3 above)
Comment (alxndrsn) [inline static/js/enketo/widgets/simprints.js]: > spaces around the function parameters
👍

File: static/js/enketo/widgets/simprints.js (hunk 5 above)
Comment (alxndrsn) [inline static/js/enketo/widgets/simprints.js]: Sounds reasonable; I hope it works :~)  From a brief play with jQuery it looks like it would:

screen shot 2017-08-14 at 12 51 09

File: translations/messages-en.properties (hunk 8 above)
Comment (alxndrsn) [inline translations/messages-en.properties]: Aha 😞

**Comment by alxndrsn** [comment]
LGTM 👍
